kernel/comps/logger/src/filter.rs

```diff
@@ -0,0 +1,145 @@
+    vec::Vec,
+};
+use core::str::FromStr;
+
+use log::{error, LevelFilter};
+
+#[derive(Debug)]
+/// `FilterEntry` is used to help the log system to do a fine-grained filtering.
+/// There are three types of `FilterEntry`:
+/// - Module type, means one want to change the [max log level] of the entire module
+///   to `log_level`.
+/// - File type, means one want to change the [max log level] of a specified file to
+///   `log_level`.
+/// - Line type, means one want to change the [max log level] of a line in a specified
+///   file to `log_level`.
+///
+/// [max log level]: The max log level within a scope ensures that any log operations
+/// with a log level greater than this value within that scope are ineffective.
+pub(crate) enum FilterEntry {
+    Module {
+        module: String,
+        log_level: LevelFilter,
+    },
+    File {
+        file: String,
+        log_level: LevelFilter,
+    },
+    Line {
+        file: String,
+        line_number: u32,
+        log_level: LevelFilter,
+    },
+}
+
+impl FilterEntry {
+    /// Creates a `FilterEntry` from an input line.
+    pub(crate) fn from_line(line: &str) -> Option<FilterEntry> {
+        let parts: Vec<&str> = line.split_whitespace().collect();
+        match parts.first() {
+            Some(&"module") => {
+                if parts.len() == 3 {
+                    Some(FilterEntry::Module {
+                        module: parts[1].to_string(),
+                        log_level: LevelFilter::from_str(parts[2]).ok()?,
+                    })
+                } else {
+                    None
+                }
+            }
+            Some(&"file") => {
+                if parts.len() == 3 {
+                    Some(FilterEntry::File {
+                        file: parts[1].to_string(),
+                        log_level: LevelFilter::from_str(parts[2]).ok()?,
+                    })
+                } else {
+                    None
+                }
+            }
+            Some(&"line") => {
+                if parts.len() == 4 {
+                    if let Ok(line_number) = parts[2].parse::<u32>() {
+                        Some(FilterEntry::Line {
+                            file: parts[1].to_string(),
+                            line_number,
+                            log_level: LevelFilter::from_str(parts[3]).ok()?,
+                        })
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+
+    pub(crate) fn log_level(&self) -> LevelFilter {
+        match self {
+            FilterEntry::Module { log_level, .. } => *log_level,
+            FilterEntry::File { log_level, .. } => *log_level,
+            FilterEntry::Line { log_level, .. } => *log_level,
+        }
+    }
+}
+
+pub(crate) fn get_filter_list() -> (Vec<FilterEntry>, LevelFilter) {
+    let file = include_str!("filter_list");
```

Comment:
The filter list is determined at the compile time, but the filter list will be visited at run time, adding significant overhead to each log statement. This is clearly a suboptimal implementation. If the filter list is determined at the compile time, why not using a proc macro to simply compile unwanted log statements to no-op? If each log output can be turned on or off at runtime, then it would be similar to [Linux tracepoints](https://lwn.net/Articles/992455/), which is implemented much more efficiently than visiting a table for each log message.